File: Educational Codeforces Round 142 (Rated for Div. 2)/C. Min Max Sort.py
t = int(input())

for _ in range(t):
    n = int(input())
    a = list(map(int, input().split()))

    # create an array p where p[i] stores the index of i in the array a
    p = [0] * (n + 1)
    for i in range(n):
        p[a[i]] = i

    # calculate the k array
    k = [0] * (n + 1)
    k[n] = 1
    for i in range(n - 1, 0, -1):
        if p[i] < p[i + 1]:
            k[i] = k[i + 1] + 1
        else:
            k[i] = 1

    # calculate the result
    res = n
    for i in range(1, n + 1):
        res = min(res, max(i - 1, n - (i + k[i] - 1)))

    print(res)

File: Educational Codeforces Round 142 (Rated for Div. 2)/C. Min Max Sort.cpp
#include<bits/stdc++.h>
using namespace std;
int main(){
    int t;
    cin>>t;
    while(t--){
        int n;
        cin>>n;
        int a[n];
        int p[n+1];
        int k[n+1];
        for(int i=0;i<n;i++){
            cin>>a[i];
            p[a[i]]=i;
        }
        k[n]=1;
        for(int i=n-1;i>=1;i--){
            if(p[i]<p[i+1]){
                k[i]=k[i+1]+1;
            }else{
                k[i]=1;
            }
        }
        int res=n;
        for(int i=1;i<=n;i++){
            res=min(res,max(i-1, n-(i+k[i]-1)));
        }
        cout<<res<<endl;
    }
}